pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSource.java

```diff
@@ -0,0 +1,169 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.io.nifi;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.nifi.remote.Transaction;
+import org.apache.nifi.remote.TransferDirection;
+import org.apache.nifi.remote.client.SiteToSiteClient;
+import org.apache.nifi.remote.client.SiteToSiteClientConfig;
+import org.apache.nifi.remote.protocol.DataPacket;
+import org.apache.nifi.stream.io.StreamUtils;
+import org.apache.pulsar.functions.api.Record;
+import org.apache.pulsar.io.core.PushSource;
+import org.apache.pulsar.io.core.SourceContext;
+import org.apache.pulsar.io.core.annotations.Connector;
+import org.apache.pulsar.io.core.annotations.IOType;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * A source that pulls data from Apache NiFi using the NiFi Site-to-Site client. This source
+ * produces NiFiDataPackets which encapsulate the content and attributes of a NiFi FlowFile.
+ */
+@Connector(
+    name = "nifi",
+    type = IOType.SOURCE,
+    help = "A simple connector to move messages from Apache NiFi using the NiFi Site-to-Site client to a Pulsar topic",
+    configClass = NiFiConfig.class)
+@Slf4j
+public class NiFiSource extends PushSource<NiFiDataPacket> {
+
+    private long waitTimeMs;
+    private volatile boolean isRunning = true;
+    private NiFiConfig niFiConfig;
+    private SiteToSiteClientConfig clientConfig;
+
+    Thread runnerThread;
+
+    @Override
+    public void open(Map<String, Object> config, SourceContext sourceContext) throws Exception {
+        niFiConfig = NiFiConfig.load(config);
+        if (niFiConfig.getUrl() == null
+                || niFiConfig.getPortName() == null) {
+            throw new IllegalArgumentException("Required property not set.");
+        }
+
+        waitTimeMs = niFiConfig.getWaitTimeMs();
+        clientConfig = new SiteToSiteClient.Builder()
+                .url(niFiConfig.getUrl())
+                .portName(niFiConfig.getPortName())
+                .requestBatchCount(niFiConfig.getRequestBatchCount())
+                .buildConfig();
+
+        this.start();
+    }
+
+    @Override
+    public void close() throws Exception {
+        this.isRunning = false;
+        if (runnerThread != null) {
+            runnerThread.interrupt();
+            runnerThread.join();
+            runnerThread = null;
+        }
+    }
+
+    protected void start() throws Exception {
+        runnerThread = new Thread(new ReceiveRunnable());
+        runnerThread.setName("Nifi Source Thread");
+        runnerThread.start();
+    }
+
+
+    class ReceiveRunnable implements Runnable {
+
+        public ReceiveRunnable() {
+        }
+
+        @Override
+        public void run() {
+            try {
+                final SiteToSiteClient client = new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
```

Comment:
nit: if `client` is a Closable, you can use `try-with-resource`.

```
try (SiteToSiteClient client = ...) {
}
```

Reply:
> nit: if `client` is a Closable, you can use `try-with-resource`.
> 
> ```
> try (SiteToSiteClient client = ...) {
> }
> ```

ok